analysis/src/python/data_analysis/templates/filter_using_diff.py

```diff
@@ -0,0 +1,134 @@
+from analysis.src.python.data_analysis.templates.utils.template_utils import parse_template_code_from_step
+from analysis.src.python.data_analysis.utils.code_utils import split_code_to_lines
+from analysis.src.python.data_analysis.utils.report_utils import parse_report
+from analysis.src.python.evaluation.tools.model.report import BaseIssue, BaseReport
+from analysis.src.python.utils.df_utils import filter_df_by_iterable_value, read_df, write_df
+from analysis.src.python.utils.logging_utils import configure_logger
+
+
+def get_issues_with_offset(report: BaseReport, code_lines: List[str]) -> List[Tuple[int, BaseIssue]]:
+    """ For each issue calculate offset in code written in one line and return as pairs sorted in ascending order."""
+
+    code_prefix_length = [0]
+    for code_line in code_lines:
+        code_prefix_length.append(code_prefix_length[-1] + len(code_line))
+
+    issues_with_offset = []
+    for issue in report.get_issues():
+        issue_offset = code_prefix_length[issue.get_line_number() - 1] + issue.get_column_number()
+        issues_with_offset.append((issue_offset, issue))
+
+    issues_with_offset.sort(key=lambda i: i[0])
+
+    return issues_with_offset
+
+
+def get_template_to_code_diffs(template_lines: List[str], code_lines: List[str]) -> List[Tuple[int, int, int]]:
+    """
+    Get template to students code diff in format of list of tuples (tag, start, end) where:
+    tag - type of change where 0 (code not changed), 1 (code added). Tag -1 (code deleted) is ignored.
+    start - position of change start (included)
+    end - position of change end (excluded)
+    """
+
+    matcher = diff_match_patch()
+    patches = matcher.diff_main(''.join(template_lines), ''.join(code_lines))
+
+    diffs = []
+    start = 0
+
+    for tag, patch in patches:
+        if tag != -1:
+            end = start + len(patch)
+            diffs.append((tag, start, end))
+            start = end
+
+    return diffs
+
+
+def get_template_issues(issues_with_offset: List[Tuple[int, BaseIssue]], diff: List[Tuple[int, int, int]]) \
+        -> List[BaseIssue]:
+    """
+    Get template issues from list of issues.
+    Issues considered as template if it's position inside change in diff with type 0 - code not changed from template.
+    """
+
+    i = 0
+    template_issues = []
+    for offset, issue in issues_with_offset:
+        while i < len(diff) and diff[i][2] < offset:
+            i += 1
+        # If tag is 0 the part of code was not changed from template
+        if i >= len(diff) or diff[i][0] == 0:
+            template_issues.append(issue)
+
+    return template_issues
+
+
+def filter_in_single_submission(submission: pd.Series, step: pd.Series, issues_column: str) -> pd.Series:
+    code_lines = split_code_to_lines(submission[SubmissionColumns.CODE.value], keep_ends=True)
+    lang = submission[SubmissionColumns.LANG.value]
+    template_lines = parse_template_code_from_step(step, lang, keep_ends=True)
```

Comment:
Should we parse it once for each step? I mean why do we parse it for each submission?

Reply:
I have done like this because I do not know if templates for different language versions are the same, but information about language version is presented only in submission. So for each submission, I need to parse it separately. But I think it takes much less time than all other parsings/processings, so it does not influence performance much.